python/ray/air/_internal/torch_utils.py

```diff
@@ -292,3 +293,221 @@ def consume_prefix_in_state_dict_if_present_not_in_place(
+                dtypes=dtypes[col_name] if isinstance(dtypes, dict) else dtypes,
+            )
+            for col_name, col_array in numpy_batch.items()
+        }
+    else:
+        numpy_list = transform_pyarrow.table_to_numpy_dict_chunked(
+            batch,
+        )
+        return convert_ndarray_list_to_torch_tensor_list(
+            numpy_list,
+            dtypes=dtypes,
+        )
+
+
+def numpy_batch_to_torch_tensors(
+    batch: Dict[str, np.ndarray],
+    dtypes: Optional[Union[torch.dtype, Dict[str, torch.dtype]]] = None,
+    device: Optional[Union[str, "torch.device"]] = None,
+) -> Dict[str, List[torch.Tensor]]:
+    """Convert a dictionary of numpy arrays to PyTorch tensors.
+
+    Args:
+        batch: Dictionary mapping column names to numpy arrays
+        dtypes: A (dict of) Torch dtype(s) for the created tensors; if None, the dtype
+            will be inferred from the NumPy ndarray data.
+        device: Optional device to place tensors on
+
+    Returns:
+        A dictionary of column name to list of tensors
+    """
+    from ray.air._internal.torch_utils import (
+        convert_ndarray_batch_to_torch_tensor_batch,
+    )
+
+    return convert_ndarray_batch_to_torch_tensor_batch(
+        batch,
+        dtypes=dtypes,
+        device=device,
+    )
+
+
+@torch.no_grad()
+def concat_tensors_to_device(
+    tensor_sequence: Sequence[torch.Tensor],
+    device: Optional[Union[str, "torch.device"]] = None,
+    non_blocking: bool = False,
+) -> torch.Tensor:
+    """Stack sequence of tensors into a contiguous GPU tensor.
+
+    Args:
+        tensor_sequence: Sequence of tensors to stack
+        device: The device to move tensors to
+        non_blocking: If True, perform device transfer without forcing a
+            synchronization.
+
+    Returns:
+        A contiguous tensor on the target device
+    """
+    # Assumes tensors have the same shape/dtype
+    assert tensor_sequence, "Cannot stack empty sequence of tensors"
+    assert all(
+        isinstance(t, torch.Tensor) for t in tensor_sequence
+    ), "All items must be torch.Tensor"
+    assert all(
+        t.dtype == tensor_sequence[0].dtype for t in tensor_sequence
+    ), "All tensors must have the same dtype"
+    assert all(
+        t.shape[1:] == tensor_sequence[0].shape[1:] for t in tensor_sequence
+    ), "All tensors must have the same shape[1:]"
+
+    first = tensor_sequence[0]
+    dtype = first.dtype
+    shape_tail = first.shape[1:]
+    total_rows = sum(t.shape[0] for t in tensor_sequence)
+
+    # Allocate an empty Tensor on device
+    result = torch.empty((total_rows, *shape_tail), dtype=dtype, device=device)
+
+    row_start = 0
+    for t in tensor_sequence:
+        row_end = row_start + t.shape[0]
+        result[row_start:row_end].copy_(t, non_blocking=non_blocking)
+        row_start = row_end
+
+    assert isinstance(result, torch.Tensor), "Result must be a torch.Tensor"
+    return result
+
+
+TensorBatchType = Union[
```

Comment:
let's move this definition to `collate_fn.py` and reference it in the doc-string when mentioning which types can be automatically transferred.